File: muxer-client/src/main/java/pt/um/tf/taskmux/client/TaskRunner.java
package pt.um.tf.taskmux.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import pt.haslab.ekit.Spread;
import pt.um.tf.taskmux.commons.messaging.URIMessage;
import pt.um.tf.taskmux.commons.task.DummyTask;
import pt.um.tf.taskmux.commons.task.Result;
import pt.um.tf.taskmux.commons.task.Task;
import spread.SpreadGroup;
import spread.SpreadMessage;

import java.util.ArrayDeque;
import java.util.Queue;
import java.util.concurrent.ExecutorService;

/**
 * The main runner of tasks.
 * Informs completion of tasks directly through to spread.
 */
public class TaskRunner {
    private final static Logger LOGGER = LoggerFactory.getLogger(TaskRunner.class);
    private Queue<Task> runningTasks;
    private ExecutorService taskPool;
    private Spread toReport;
    private SpreadGroup maingroup;
    private int barrier;

    public TaskRunner(ExecutorService taskPool,
                      int barrier,
                      Spread toReport) {
        runningTasks = new ArrayDeque<>(barrier);
        this.taskPool = taskPool;
        this.maingroup = null;
        this.toReport = toReport;
        this.barrier = barrier;
    }

    /**
     * Joins aren't immediately effective.
     * @param maingroup
     */
    public void setMaingroup(SpreadGroup maingroup) {
        this.maingroup = maingroup;
    }

    public boolean runTask(Task t) {
        var res = false;
        if (!isBackedUp()) {
            if (t instanceof DummyTask) {
                var dt = (DummyTask) t;
                dt.setExecutor(taskPool);
                dt.start()
                  .thenAcceptAsync(r -> handler(dt, r), taskPool)
                  .exceptionally(th -> {
                      LOGGER.error("", th);
                      runningTasks.remove();
                      return null;
                  });
                runningTasks.add(t);
                res = true;
            }
            else {
                LOGGER.error("Unknown task type");
            }
        }
        return res;
    }

    private void handler(DummyTask dt, Result<Long> longResult) {
        //Result doesn't really matter.
        if(!longResult.completedSuccessfully()) {
            LOGGER.error("Failed task :", longResult.completedWithException());
        }
        LOGGER.info("Completed task : " + longResult.completeWithResult());
        runningTasks.remove();
        sendURLMessage(dt);
    }

    private void sendURLMessage(Task ran) {
        var spm = new SpreadMessage();
        spm.addGroup(maingroup);
        spm.setSafe();
        toReport.multicast(spm, new URIMessage(ran.getURI()));
    }

    public boolean isBackedUp() {
        return runningTasks.size() >= barrier;
    }

}

File: muxer-client/src/main/kotlin/pt/um/tf/taskmux/client/TaskRunner.kt
package pt.um.tf.taskmux.client

import mu.KLogging
import pt.haslab.ekit.Spread
import pt.um.tf.taskmux.commons.messaging.URIMessage
import pt.um.tf.taskmux.commons.task.DummyTask
import pt.um.tf.taskmux.commons.task.Result
import pt.um.tf.taskmux.commons.task.Task
import spread.SpreadGroup
import spread.SpreadMessage
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.function.Consumer

/**
 * The main runner of tasks.
 * Informs completion of tasks directly through to spread.
 */
class TaskRunner(private val taskPool : ExecutorService,
                 private val barrier : Int,
                 private val spread : Spread) {
    companion object : KLogging()
    private val runningTasks : Queue<Task<out Any>> = ArrayDeque(barrier)
    private var maingroup : SpreadGroup? = null

    /**
     * Joins aren't immediately effective.
     * @param maingroup
     */
    fun setMaingroup(maingroup : SpreadGroup) {
        this.maingroup = maingroup
    }

    fun runTask(t : Task<out Any>) : Boolean {
        var res = false
        if (!isBackedUp()) {
            if (t is DummyTask) {
                t.executor = taskPool
                t.start()
                 .thenAcceptAsync(this.accept(t), taskPool)
                 .exceptionally {
                     logger.error("", it)
                     runningTasks.remove()
                     return@exceptionally null
                 }
                runningTasks.add(t)
                res = true
            }
            else {
                logger.error("Unknown task type")
            }
        }
        return res
    }

    private fun accept(dt: DummyTask) : Consumer<Result<Long>> {
        return Consumer {
            handler(dt, it)
        }
    }

    private fun handler(dt : DummyTask, longResult : Result<Long>) {
        //Result doesn't really matter.
        if(!longResult.completedSuccessfully()) {
            logger.error("Failed task :", longResult.completedWithException())
        }
        logger.info("Completed task : " + longResult.completeWithResult())
        runningTasks.remove()
        sendURLMessage(dt)
    }

    private fun sendURLMessage(ran : Task<out Any>) {
        var spm = SpreadMessage()
        spm.addGroup(maingroup)
        spm.setSafe()
        spread.multicast(spm, URIMessage(ran.uri))
    }

    fun isBackedUp() : Boolean {
        return runningTasks.size >= barrier
    }

}
